packages/react/roving-focus/src/RovingFocusGroup.tsx

```diff
@@ -165,6 +165,11 @@ function getRovingFocusItems(groupId: string): HTMLElement[] {
   return Array.from(document.querySelectorAll(`[${GROUP_ID_DATA_ATTR}="${groupId}"]`));
 }
 
+// items that are not visible (ie: display: none) should not be cycled through
+function filterVisibleItems(elements: HTMLElement[]) {
+  return elements.filter((item) => Boolean(item.offsetParent));
```

Comment:
My initial thoughts here are this is probably not a good idea as reading `.offsetParent` will cause layout thrashing every single time.

Have you been able to identify what actually causes this issue in our case?

When I test other libraries, they seem to be able to get past the hidden one (allbeit with an extra keydown).

Reply:
that's a good point regarding `.offsetParent`. Since this is done as a consequence of a user interaction (arrow press), I think it should be fine performance wise, but I understand your reticence in adding this considering this is a low-level utility.
I can try and investigate what other libraries are doing, to see if I find a better solution

Reply:
I'm going to look at it on my side too and try to first understand what is happening on our side.

Reply:
Ok I think I know what's happening. Here are my findings:

- browsers won't focus something that is not visible, ie. `display: none`, `visibility: hidden`, `hidden` attribute, (however they will focus `opacity: 0`)
- as far as I can tell this results in the browser not moving the focus at all from where it currently is
- because our roving focus solution is always based on the `document.activeElement` as the "current" one it won't let us go further
- the reason reakit, react-aria, etc keep working (allbeit with the extra keydown, they have the same issue basically) is probably because they use an index to know what the current one is
- I did a quick test with an index and it does work the same way as theirs, meaning that when you keydown and the next item is not visible, it won't move, but the index is incremented, so the next keydown will jump over it this time

I think the issue actually probably goes beyond roving focus but is rather a general focus issue (ie. the same issues probably exist with our focus trapping solution for example).

I'm not too sure yet if/how we should address this. Accurately knowing when items are visible is not super cheap so I'm little wary about it. I'll discuss this with @jjenzz and we'll see what we can do.

Hopefully in the meantime you can get around it using the alternative I proposed here: https://github.com/radix-ui/primitives/issues/581#issuecomment-811796630

Reply:
thanks for checking!